substrate/primitives/core/src/bandersnatch.rs

```diff
@@ -774,51 +762,45 @@ pub mod ring_vrf {
 		///
 		/// The signature is verifiable if it has been produced by a member of the ring
 		/// from which the [`RingVerifier`] has been constructed.
-		pub fn verify(&self, data: &VrfSignData, verifier: &RingVerifier) -> bool {
+		pub fn ring_vrf_verify(&self, data: &VrfSignData, verifier: &RingVerifier) -> bool {
 			const _: () = assert!(MAX_VRF_IOS == 3, "`MAX_VRF_IOS` expected to be 3");
 			let preouts_len = self.outputs.len();
 			if preouts_len != data.inputs.len() {
 				return false
 			}
 			// Workaround to overcome backend signature generic over the number of IOs.
 			match preouts_len {
-				0 => self.verify_gen::<0>(data, verifier),
-				1 => self.verify_gen::<1>(data, verifier),
-				2 => self.verify_gen::<2>(data, verifier),
-				3 => self.verify_gen::<3>(data, verifier),
+				0 => self.ring_vrf_verify_gen::<0>(data, verifier),
+				1 => self.ring_vrf_verify_gen::<1>(data, verifier),
+				2 => self.ring_vrf_verify_gen::<2>(data, verifier),
+				3 => self.ring_vrf_verify_gen::<3>(data, verifier),
 				_ => unreachable!(),
 			}
 		}
 
-		fn verify_gen<const N: usize>(&self, data: &VrfSignData, verifier: &RingVerifier) -> bool {
-			let Ok(preoutputs) = self
-				.outputs
-				.iter()
-				.map(|o| o.0.clone())
-				.collect::<arrayvec::ArrayVec<bandersnatch_vrfs::VrfPreOut, N>>()
-				.into_inner()
-			else {
-				return false
-			};
-
-			let Ok(signature) =
-				PedersenVrfSignature::deserialize_compressed(self.signature.as_slice())
+		fn ring_vrf_verify_gen<const N: usize>(
+			&self,
+			data: &VrfSignData,
+			verifier: &RingVerifier,
+		) -> bool {
+			let Ok(vrf_signature) =
+				bandersnatch_vrfs::RingVrfSignature::<0>::deserialize_compressed(
+					self.signature.as_slice(),
+				)
 			else {
 				return false
 			};
 
-			let Ok(ring_proof) = RingProof::deserialize_compressed(self.ring_proof.as_slice())
-			else {
-				return false
-			};
+			let preouts: [bandersnatch_vrfs::VrfPreOut; N] =
+				core::array::from_fn(|i| self.outputs[i].0.clone());
```

Comment:
How big is the thing being cloned here? I tried to follow the types and manually see how big it is, but there are so many layers of abstraction between what's used here and the types which actually store the raw bytes (and some broken `docs.rs` builds in between :P) that I just gave up. (As far as I can see this has some inline big ints inside, so I'd like to know how big exactly.) Can you `core::mem::size_of` in a throwaway `#[test]` and paste me the result?

Reply:
Each element is a Bandersnatch curve group element (aka curve point).
```
struct Affine {
    x: U256,            // 32 bytes
    y: U256,            // 32 bytes
    infinity: bool,
}
```

The printed size is 72 bytes (8 bytes are given to the `infinity` bool for alignment).

---

The constructed array size is bounded by the max number we support at the moment (`MAX_VRF_IOS = 3`).

Reply:
@burdges probably we can implement `Copy` for `VrfInput` and `VrfPreOut`   (as the inner types are Copy)

Reply:
Alright sure.  Arkworks & dalek both have copy types of this size.

In memory, yeah affine curve points are two base field elements plus a flag.  This one has a 32 byte base field, and serializes in 33 bytes.

Yes, arkworks has layers itself.  And I made it worse for other reasons.